com.unity.netcode.gameobjects/Runtime/Serialization/IBufferSerializerImplementation.cs

```diff
@@ -0,0 +1,37 @@
+using System;
+using UnityEngine;
+
+namespace Unity.Netcode
+{
+    public interface IBufferSerializerImplementation
+    {
+        bool IsReader { get; }
+        bool IsWriter { get; }
+
+        ref FastBufferReader GetFastBufferReader();
+        ref FastBufferWriter GetFastBufferWriter();
+
+        void SerializeValue(ref object value, Type type, bool isNullable = false);
+        void SerializeValue(ref GameObject value);
+        void SerializeValue(ref NetworkObject value);
+        void SerializeValue(ref NetworkBehaviour value);
+        void SerializeValue(ref string s, bool oneByteChars = false);
```

Comment:
huh? what's that optional param? if this is needed for encoding, I'd suggest passing `System.Text.Encoding` stuff around so that we could serialize string <-> byte[] back & forth properly.

Reply:
The optional param came from the old NetworkReader/NetworkWriter implementations. I was trying to support everything they support. I could remove it from the Serializer, but it's there on Reader/Writer so I figured it didn't hurt to put it on the Serializer too.

Reply:
is this to say use 1 byte per 1 char, which would mean an optimization for ASCII encoding?

Reply:
Yes.

Reply:
let's either put Encoding onto API or kill it and use UTF8 internally to convert between string <-> byte[]?

Reply:
I'd like to get other opinions on this one before I make that change, since the decision to do it this way predates this PR and changing that isn't something I feel should be done without discussion... @mattwalsh-unity @LukeStampfli ?

Reply:
I don't like the API but yeah let's address this separately as it's not related to this.

Reply:
I think we should not allow variable length strings anyway. They are just horrible; people do bad stuff with them. We have fixed string classes that make it less messy and easy to put reasonable limits.

Reply:
I agree, but we're not in a position right now to get rid of them. We still require variable length things (strings and otherwise) in multiple places in the SDK core and I neither have the time nor the authority to clean that up for release... so we're stuck having to support them for now. Which is sad.

Reply:
> let's address this separately as it's not related to this.

we're literally looking at the interface and a PR that does a whole bunch of serialization overhaul.
I think this conversation is quite relevant and if we can decide on an approach, it needs to be done here too.

Reply:
I personally think we should have the Encoding passed through the interface so that we can use that behind the scenes when we do string <-> byte[] converstion and also people can offer Encoding.UTF8 or Encoding.ASCII as a default param if they want.

Reply:
I honestly don't have a dog in this fight. I have no opinion on this parameter. That's why I'm asking opinions from others, because I'm not the originator of the parameter and was hoping whoever was would weigh in.

Reply:
My only opinion is that we shouldn't ever use this at all. In fact I'm pretty sure we currently don't serialize strings anywhere except for ServerLogMessage... could we just kill this entirely from the interface?